runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java

```diff
@@ -70,7 +72,7 @@ protected void beforeExecute(Thread t, Runnable r) {
           protected void afterExecute(Runnable r, Throwable t) {
             super.afterExecute(r, t);
             synchronized (this) {
-              if (activeCount.getAndDecrement() == maximumPoolSize) {
+              if (activeCount.getAndDecrement() == maximumThreadCount.get()) {
```

Comment:
I think this might need to be
if (activeCount.getAndDecrement() <= maximumThreadCount.get() &&
    startTimeMaxActiveThreadsUsed > 0)

you could maybe test this by adding more work to queue than the limit, having the work that is scheduled block, then increase the limit and let all the work complete.  We should see totalTimeMaxActiveThreadsUsed incremented

Another way to fix it could be to modify the setMaximumPoolSize to reset and count startTimeMaxActiveThreadsUsed if the limit increased (it will be restarted shortly as new threads start and call beforeExecute).

Reply:
Yeah that makes sense.